doc/modules/array_api.rst

```diff
@@ -178,21 +178,82 @@ Tools
 
-Type of return values and fitted attributes
--------------------------------------------
+Input and output array type handling
+------------------------------------
 
-When calling functions or methods with Array API compatible inputs, the
+Estimators and scoring functions are able to accept input arrays
+from different array libraries and/or devices. When mixed input arrays are
+passed, scikit-learn silently converted to make them all consistent.
+
+For estimators, the rule is **"everything follows `X`"** - mixed array inputs are
+converted so they are all match the array library and device of `X`.
+For scoring functions the rule is **"everything follows `y_pred`"** - mixed array
+inputs are converted so they are all match the array library and device of `y_pred`.
+
+When a function or method has been called with Array API compatible inputs, the
 convention is to return array values of the same array container type and
 device as the input data.
 
-Similarly, when an estimator is fitted with Array API compatible inputs, the
-fitted attributes will be arrays from the same library as the input and stored
-on the same device. The `predict` and `transform` method subsequently expect
+For details and reasons behind this choice, see below.
+
+Estimators
+~~~~~~~~~~
+
+When an estimator is fitted with an Array API compatible `X`, all other
+array inputs, including constructor arguments, (e.g., `y`, `sample_weight`,
+`weights_init`) will be silently converted to match the array library and
+device of `X`, if they do not already.
+This allows estimators to accept mixed input types, enabling `X` to be moved
+to a different device within a pipeline, without explicitly moving `y`.
+Note that scikit-learn pipelines do not allow transformation of `y` (to avoid
+:ref:`leakage <data_leakage>`).
+
+Take for example a pipeline where `X` and `y` both start on CPU, and go through
+the following three steps:
+
+* :class:`~sklearn.preprocessing.TargetEncoder`, which will transform categorial
+  `X` but also requires `y`, meaning both `X` and `y` need to be on CPU.
+* :class:`~sklearn.preprocessing.FunctionTransformer`, with the function
+  `partial(torch.asarray, device="cuda")`, which moves `X` to GPU, to improve
+  performance in the next step.
+* :class:`~sklearn.linear_model.Ridge`, whose performance can be improved when
+  passed arrays on GPU, as it only takes numerical inputs.
+
+`X` initially contains categorical string data (thus needs to be on CPU), which is
+target encoded to numerical values in :class:`~sklearn.preprocessing.TargetEncoder`.
+`X` is then explicitly moved to GPU to improve the performance of
+:class:`~sklearn.linear_model.Ridge`. `y` cannot be transformed by the pipeline
+(recall scikit-learn pipelines do not allow transformation of `y`) but as
+:class:`~sklearn.linear_model.Ridge` is able to accept mixed input types,
+this is not a problem and the pipeline is able to be run.
```

Comment:
> At first sight it seems that the y follows X principle here is broken and upon thinking about it, I could then see that this part is to demonstrate how y being on the different namespace on the beginning of a new step of a pipeline demonstrates exactly how it follows X without that the user needs to take care of it. But right now, there is only a very subtle hint on this in this paragraph and this idea is not clearly expressed.

@StefanieSenger this is the only part that explicit talks about this. I tried to make it clearer but open to suggests to improve. Thanks!